### ISSUE TITLE
The result of feature parallel algorithm is not correct

### ISSUE THREAD

**Issue by songchuangyuan** [microsoft/LightGBM#3292]
1. Directly use the stand-alone version of the binary_classification example test, the number of rows of data is 7000, the feature is 28, set the training parameters, calculate the AUC and loss value.
2. Using the example test of two parallel machines, the data in the first step is divided vertically. The number of data rows of the first machine is 7000 and feature 14, and the number of data rows of the second machine is 7000 and feature 14. Set the same parameters as the first step, calculate the AUC and loss values.

Result: Find that the results are not consistent, and there is a big difference. if you want to do parallelism, you should get the same result as the two steps. Why not consistent?

**Comment by guolinke** [comment]
Some settings will break the consistency of parallel algorithm.
could you provide your settings?

**Comment by guolinke** [comment]
BTW, for the feature palallel in lightgbm, you need to pass the whole dataset (with 28 featues), to both two nodes.

**Comment by songchuangyuan** [comment]
> BTW, for the feature palallel in lightgbm, you need to pass the whole dataset (with 28 featues), to both two nodes.

Thank you for your answer. I tested it again according to your method, and the result is correct. In fact, the feature parallel mode is only to speed up the training, not because the file is too large to load into the memory. What about data parallelism? 

I did the first test:
1. Directly use the stand-alone version of the binary_classification example test, the number of rows of data is 7000, the feature is 28, set the training parameters, calculate the AUC and loss value.
2. Using the example test of two parallel machines, the data in the first step is divided horizontally. The number of data rows of the first machine is 3500 and feature 28, and the number of data rows of the second machine is 3500 and feature 28. Set the same parameters as the first step, calculate the AUC and loss values.
Result: It is found that the results of the two steps are not consistent..

I did the second test:
1. Directly use the stand-alone version of the binary_classification example test, the number of rows of data is 7000, the feature is 28, set the training parameters, calculate the AUC and loss value.
2. Using the example test of two parallel machines, both machines have the complete data from the first step. Set the same parameters as the first step, calculate the AUC and loss values.
Result: It is found that the results of the two steps are still inconsistent, and the results on the two machines are also inconsistent.

So how is the data parallel mode implemented?
It also needs to have the whole dataset(7000 rows) on both two nodes, but each node only reads a part of the dataset, create bins and construct histograms respectively, and then aggregates the histogram generated on different nodes using reduce scatter? In this step of creating bins, do you only need to use part of the local dataset or the whole dataset?

In addition, the data parallel test configuration information is as follows:
task = train
boosting_type = gbdt
objective = binary
metric = binary_logloss,auc
metric_freq = 1
is_training_metric = true
max_bin = 255
data = binary.train
valid_data = binary.test
num_trees = 100
learning_rate = 0.1
num_leaves = 15
tree_learner = data
feature_fraction = 1
bagging_freq = 5
bagging_fraction = 1
min_data_in_leaf = 5
min_sum_hessian_in_leaf = 5.0
is_enable_sparse = true
use_two_round_loading = false
is_save_binary_file = false
output_model = LightGBM_model.txt
num_machines = 2
local_listen_port = 12400
machine_list_file = mlist.txt

**Comment by guolinke** [comment]
@songchuangyuan 
1. if you don't set `pre_prationition=true`, you should use the whole dataset for all machines.
2. The bin-finding (feature value bucketize) is also parallelize. Each machine will only use its local data, to find the bins for a subset of features, just like your observation. The reason is for efficiency since it is not efficient to sync the samples in different machines (we assuming each node only have part of samples). Moreover, you cannot only use the local data for bin-finding of all features, otherwise the bucket ranges will be different in different machines. 
3. Due to 2, The data parallelism may is not consistent with serial algorithm, due to the feature bin bucket ranges may is different. But in the very large scale data, it should be very close.

**Comment by songchuangyuan** [comment]
> @songchuangyuan
> 
> 1. if you don't set `pre_partition=true`, you should use the whole dataset for all machines.
> 2. The bin-finding (feature value bucketize) is also parallelize. Each machine will only use its local data, to find the bins for a subset of features, just like your observation. The reason is for efficiency since it is not efficient to sync the samples in different machines (we assuming each node only have part of samples). Moreover, you cannot only use the local data for bin-finding of all features, otherwise the bucket ranges will be different in different machines.
> 3. Due to 2, The data parallelism may is not consistent with serial algorithm, due to the feature bin bucket ranges may is different. But in the very large scale data, it should be very close.

@guolinke 
Thank you very much for your reply. It is helpful to understand the parallel algorithm of lightgbm. I'm studying the federal machine learning of homomorphism data based on lightgbm. According to your answer, the problem to be solved is still different from that of data parallel thinking.
